### ISSUE TITLE
BOHB may try duplicate hyper-parameters in the same SH round with the same budget

### ISSUE THREAD

**Issue by PurityFan** [automl/HpBandSter#53]
When I use BOHB to optimize hyperparameter tuning, I found that in the same SH round with the same budget, it may have same hyper-parameter configurations.

For example, here is the parameter that chosen by my experiment using BOHB:

{"optimizer": "Adagrad", "model": "mobilenet", "lr": 0.0001, "budget": 81},
{"optimizer": "Adagrad", "model": "mobilenet", "lr": 0.0001, "budget": 81}, 
{"optimizer": "Adagrad", "model": "dpn92", "lr": 0.001, "budget": 81}, 
{"optimizer": "Adagrad", "model": "mobilenet", "lr": 0.0001, "budget": 81}, 
{"optimizer": "Adadelta", "model": "dpn92", "lr": 0.0001, "budget": 81}, 

Actually, {"optimizer": "Adagrad", "model": "mobilenet", "lr": 0.0001} doesn't performing well, therefore, repeated selection wastes lots of computing resources, is there any ways to help me avoid this situation?

Thanks so much!

**Comment by sfalkner** [comment]
The values of the learning rate and the other parameters suggest that your configuration space contains only 3 categorical variables, is that right?
If that is the case, BOHB is definitely the wrong optimizer as it does not actively tries to keep the evaluated configurations different.  Could you please provide some more detail on your search space, the budgets and the number of iterations you want to run? That way I might be able to give you some better advice.

**Comment by PurityFan** [comment]
Thanks for the quick response.

The search space:
{
    "lr":{(choice): [0.1, 0.01, 0.001, 0.0001]},
    "optimizer":{(choice): ["SGD", "Adadelta", "Adagrad", "Adam", "Adamax"]},
    "model":{(choice): ["vgg", "resnet18", "googlenet", "densenet121", "mobilenet", "dpn92", "senet18"]}
}
I set the max_budget=243, the min_budget=1, eta = 3, others as default.

I have checked the hyperparameter sets generated in the first SH round, and its distribution is random and uniform. When it comes to the 81 budget round, the same parameters appear frequently?

**Comment by sfalkner** [comment]
We know BOHB struggles with purely discrete spaces. Your's has only a 140 different configurations, and there is no mechanism in BOHB that prevents it from reevaluating the same configurations over and over again (except for the fraction of random configurations). I assume that the config you mention:
 {"optimizer": "Adagrad", "model": "mobilenet", "lr": 0.0001}
works well on the smaller budgets which is why BOHB's model will be biased very early and will become overly certain about that.

You can try to make the learning rate a continuous parameter which would probably already fix the problem.

BTW:
1. as your space has only 140 configurations, the first iteration (which tries out 243 configurations with budget=1)  already evaluates certain configurations multiple times.
2. I don't know what kind of experiments you are running, but you might want to include some learning rate schedule for your training which usually also comes with parameters that need tuning. That usually boosts performance.

**Comment by PurityFan** [comment]
Yes, although there are only 140 cases in the search space, it is not reasonable to have repeated parameters because this repeated hyperparameter does not perform well (by observing previous performance).

I grid search 140 different configuration, and I set b_max as 243 and b_min as 1(so it will try 243 configurations at the first SH). But I found that the best performance hyperparameters **haven't been tried** in BOHB.

So my point is whether BOHB **supports the check mechanism**. BOHB may have bias for some parameters. But if this hyperparameter has been selected in this SH round, can BOHB put the resources on other hyperparameters (although it's previously performed is not so good, is it better than doing repeated hyperparameter trials?)

**Comment by sfalkner** [comment]
BOHB will evaluate 243 random configurations on the smallest budget. There is a good chance that not all of your 140 combinations are tried, because there is no mechanism to prevent evaluating the same configuration twice. That being said, of those 243, the best 81 will be evaluated on budget 3. The best 27 of these on budget 9, and so on. So the first iteration is heavily biased towards the configuration that perform well for small budgets. By the time for the second iteration, the build model inside BOHB is probably stuck in a local optima and is overly confident that it has found the best config.
For your space there is not much you can do other than increasing `b_min` to say 27 and maybe try setting `min_bandwidth=1e-1` which strongly encourages exploration.
Your space is a rather extreme one because it has a rather small number of discrete configurations in it.

**Comment by PurityFan** [comment]
I see, thank you very much. And also thank you and your team for the contribution to the algorithm of autoML, outstanding work!
